src/solver/optim-model-filler/include/antares/solver/optim-model-filler/VariableDictionary.h

```diff
@@ -140,60 +163,113 @@ class Dimensions
+    }
+    return ret;
+}
+
+inline std::string buildVariableName(const PartialKey& key, MCYearAndTime mcyearAndTime)
+{
+    std::string ret = fmt::format("{}.{}", key.getComponent(), key.getVariable());
+    ret += "_s" + std::to_string(format_as(mcyearAndTime.mcYear));
+    ret += "_t" + std::to_string(mcyearAndTime.timestep);
+    return ret;
+}
+
 /**
  * @brief Storage for created MIP variables indexed by (component, variable, scenario?, timestep?).
  * Uses nested hash maps + a vector with offset for contiguous timestep indices.
  * Provides strict (throwing) accessors and non-throwing tryGet helpers.
+ * Optimized version with O(1) access by index for hot path performance.
  */
 template<Antares::Optimisation::LinearProblemApi::SolverVariable InnerSolverVariable>
 class VariableDictionary
 {
     using Value = Antares::Optimisation::LinearProblemApi::IMipVariable<
       InnerSolverVariable>*; ///< Pointer alias
 
+    // Legacy storage for backward compatibility
     class VectorWithOffset
     {
     public:
         VectorWithOffset() = default;
-        /** @brief Resize underlying storage; new slots initialised to nullptr. */
         void resize(size_t initial_size, unsigned offset);
-        Value& operator[](unsigned int index); ///< Unchecked access (existing)
-        [[nodiscard]] const Value& operator[](unsigned int index) const; ///< Unchecked const
-        [[nodiscard]] const Value& at(unsigned int index) const;         ///< Bounds-checked const
-        Value& at(unsigned int index);                                   ///< Bounds-checked
-        [[nodiscard]] bool contains(unsigned int index) const noexcept;  ///< Fast bounds test
+        Value& operator[](unsigned int index);
+        [[nodiscard]] const Value& operator[](unsigned int index) const;
+        [[nodiscard]] const Value& at(unsigned int index) const;
+        Value& at(unsigned int index);
+        [[nodiscard]] bool contains(unsigned int index) const noexcept;
 
     private:
         std::vector<Value> values_ = {};
-        unsigned int offset_ = 0; ///< Logical start index used for mapping
+        unsigned int offset_ = 0;
     };
 
     using TwoIndexVectorByYear = std::unordered_map<MCYearAndTime::MCYear, VectorWithOffset>;
     using HashMapVector = std::unordered_map<PartialKey, TwoIndexVectorByYear, PartialKeyHash>;
 
-    HashMapVector storageOfAddedMipVariables_;                         ///< Main storage
-    const TwoIndexVectorByYear& operator[](const PartialKey& k) const; ///< Internal helper
-    mutable StringToIdMapper mapper_;                                  ///< For naming
+    // Legacy storage (kept for compatibility)
+    HashMapVector storageOfAddedMipVariables_;
+    // New optimized storage
+    std::vector<Value> storage_; ///< O(1) access by index
+    std::unordered_map<PartialKey, KeyHandle, PartialKeyHash>
+      keyHandleMap_; ///< PartialKey -> Handle resolution
+
+    mutable StringToIdMapper mapper_; ///< For naming
```

Comment:
For simplicity, I suggest keeping a small number of containers, instead of the 3 presented here.